collect_app/src/main/java/org/odk/collect/android/formentry/ODKView.java

```diff
@@ -475,40 +483,70 @@ public void scrollTo(@Nullable QuestionWidget qw) {
+
+        if (bundle != null) {
+            Set<String> keys = bundle.keySet();
+            for (String key : keys) {
+                for (QuestionWidget questionWidget : widgets) {
+                    FormEntryPrompt prompt = questionWidget.getFormEntryPrompt();
+                    TreeReference treeReference =
+                            (TreeReference) prompt.getFormElement().getBind().getReference();
+
+                    if (treeReference.getNameLast().equals(key)) {
+                        switch (prompt.getDataType()) {
+                            case Constants.DATATYPE_TEXT:
+                                formController.saveAnswer(prompt.getIndex(),
+                                        ExternalAppsUtils.asStringData(bundle.get(key)));
+                                break;
+                            case Constants.DATATYPE_INTEGER:
+                                formController.saveAnswer(prompt.getIndex(),
+                                        ExternalAppsUtils.asIntegerData(bundle.get(key)));
+                                break;
+                            case Constants.DATATYPE_DECIMAL:
+                                formController.saveAnswer(prompt.getIndex(),
+                                        ExternalAppsUtils.asDecimalData(bundle.get(key)));
+                                break;
+                            default:
+                                throw new RuntimeException(
+                                        getContext().getString(R.string.ext_assign_value_error,
+                                                treeReference.toString(false)));
+                        }
+
+                        ((StringWidget) questionWidget).setDisplayValueFromModel();
+                        break;
                     }
+                }
+            }
+        }
+
+        if (clipData != null) {
+            for (int i = 0; i < clipData.getItemCount(); i++) {
+                CharSequence key = clipData.getItemAt(i).getText();
+                if (key == null) {
+                    continue;
+                }
 
-                    ((StringWidget) questionWidget).setDisplayValueFromModel();
-                    break;
+                for (QuestionWidget questionWidget : widgets) {
+                    FormEntryPrompt prompt = questionWidget.getFormEntryPrompt();
+
+                    if (FormEntryPromptUtils.getQuestionName(prompt).equals(key.toString())
+                            && prompt.getDataType() == Constants.DATATYPE_BINARY) {
+                        Uri uri = clipData.getItemAt(i).getUri();
+                        if (uri != null) {
+                            try {
+                                File destFile = FileUtils.createDestinationMediaFile(formController.getInstanceFile().getParent(), ContentResolverHelper.getFileExtensionFromUri(getContext(), uri));
+                                FileUtils.saveMediaFileFromUri(uri, destFile, getContext());
```

Comment:
This could maybe use the `QuestionMediaManager` in some way. I've ended up playing with that in #4190 so might be better to come back fro that later!

Reply:
That might be a good idea, I've added a TODO comment to improve it later.